index.js

```diff
@@ -448,6 +448,54 @@ const publicApi = {
     },
 
     /**
+     * If enable, browser sync will start to provide live reload functionality
+     *
+     * https://www.browsersync.io/
+     *
+     * Call it to start a proxy redirecting requests to `backendUrl` while
+     * watching `paths` for changes. It start when using any of:
+     *
+     *      encore dev --watch
+     *      encore dev-server
+     *
+     * Use `browserSyncOptionsCallback` callback to configure `browser-sync`
+     * and `browser-sync-webpack-plugin` options.
+     *
+     * Options configured in callback can be used to overwrite everything,
+     * included `backendUrl` and `paths`. Notice that adding them as parameter
+     * here is just an idiomatic sugar to make easier providing sensible
+     * defaults and/or calling it.
+     *
+     *     Encore.enableBrowserSync(
+     *         'http://localhost:3100/',
+     *         ['./templates/*.twig', './src/*.php']
+     *     );
+     *
+     *     // or configure its options
+     *     // https://www.browsersync.io/docs/options
+     *     // https://github.com/Va1/browser-sync-webpack-plugin
+     *     Encore.enableBrowserSync('http://localhost:3100/',
+     *         ['./*.twig', './*.php']
+     *         function(browserSyncOptions, pluginOptions) {
+     *             browserSyncOptions.host = 'localhost';
+     *             // overrides `backendUrl` parameter
+     *             browserSyncOptions.proxy = 'http://localhost:8080'
+     *             pluginOptions.reload= false;
+     *         }
+     *     );
+     *
+     * @param {string} backendUrl
+     * @param {Array} paths
+     * @param {function} browserSyncOptionsCallback
+     * @return {publicApi}
+     */
+    enableBrowserSync(backendUrl, paths, browserSyncOptionsCallback = () => {}) {
```

Comment:
I'm wondering if it makes sense to have the `backendUrl` argument there. Couldn't we provide a default value for proxy/port and let people that need to override it use the `browserSyncOptionsCallback`?

Kind of the same remark about `paths` but I understand the need for that one a bit better.

Reply:
Hi @Lyrkan,

it was just idiomatic thing like I said, I just thought that most use cases fit on this one, from my point of view it makes method signature to look nicer. We can change it, of course, although would be nice to think how will look method signature with `paths` as first argument only like this:  
```
Encore.enableBrowserSync(['./*.twig', './*.php'], function(browserSyncOptions, pluginOptions) {
    browserSyncOptions.proxy = 'http://localhost:8080';
    browserSyncOptions.port = '8080';
});
```
instead this:
```
Encore.enableBrowserSync(
    'http://localhost:8080', 
    ['./*.twig', './*.php'],  
    function(browserSyncOptions, pluginOptions) {}
);
```
Leaving only `paths` make me think we should remove it also and specify them in callback, or not?
Then we will have more boilerplate code in our callback always :(

some decision must be made regarding this one.

Reply:
The questions we have to ask ourselves are:

* will most users actually need to change the value of proxy and port?
* will it be more frequent than having to modify paths?
* will it also be more frequent than having to modify other options through the callback?

The same questions have to be asked for `paths`.

> Then we will have more boilerplate code in our callback always :(

I'm not sure what you mean by that, can't we provide a sensible default value for each of them in `browser-sync.js` ?

Reply:
> will most users actually need to change the value of proxy and port?

Scenarios I have found are:
1. Usage of php built-in server. This I use heavily. Is normal to have couple of project running at same time with different ports.
2. Docker setup. Many colleagues use a docker container. In there, server is under vhost on default port 80/443 or just any port they choose to redirect to/from container, mainly because they also run several container at same time too.
3. Default server install (nginx/apache) with vhosts and port 80/443.

> will it be more frequent than having to modify paths?

Regarding frequent changes on setup, honestly I don't know how could be, from my experience, it will change, even if the change it is not in repository committed, there will be cases for example where changing port for developing will happen.

> I'm not sure what you mean by that, can't we provide a sensible default value for each of them in browser-sync.js ?

We can provide them, I did already when taking care of moving the variable initialization we talked before. But at the moment someone needs something different, which has been proven is happening most of the cases, then they will be helpless because of lack of configuration.

Actually discussing these ideas I have come to think that best compromise to avoid boilerplate code and to have still some flexibility is to use not the signature proposed initially but the one proposed below:

```
Encore.enableBrowserSync([prox url], [port], [paths to watch], [options callback]);
```

Leaving everything to configure in callback we would need to configure every time the same as we have now, only variations would be done in **[proxy url]**, **[port]**, and **[paths to watch]***, rest would remain the same. The options we have now are:

```
    {
        proxy: [proxy url],
        port: [port],
        files: [
            {
                match: [path to watch],
                fn: function(event, file) {
                    if (event === 'change') {
                        // get the named instance
                        const bs = require('browser-sync').get('bs-webpack-plugin');
                        bs.reload();
                    }
                }
            }
        ]
    }
```
that is what I meant as boilerplate code, `files` array options, with a callback inside to execute on `change` event. All that will be the same always.

Reply:
> 1. Usage of php built-in server. This I use heavily. Is normal to have couple of project running at same time with different ports.
> 2. Docker setup. Many colleagues use a docker container. In there, server is under vhost on default port 80/443 or just any port they choose to redirect to/from container, mainly because they also run several container at same time too.
> 3. Default server install (nginx/apache) with vhosts and port 80/443.

For the port these 3 cases should not be an issue if we avoid 80/443/8080/8443/8000 in our default values.

The proxy URL would not be an issue either since it'll nearly always be `http://localhost:<dev server port>` (but that implies that we can actually retrieve the port used by the dev server, see my other comment below).

> But at the moment someone needs something different, which has been proven is happening most of the cases, then they will be helpless because of lack of configuration.

I agree that paths are a bit more complicated so maybe we should actually leave that one in parameters, but I'm not so sure about the proxy url and port:

```js
Encore.enableBrowserSync([/* paths */], (bsOptions) => {
  bsOptions.port = 3002;
});
```